### PR TITLE
Dump unique NFTs from a signed state file

### LINKED ISSUE #8255: Differential Testing Analytical Engine: Dump unique NFTs, as text, from state file
Dump unique NFTs as text from a state file. Deterministically, in the sense that the NFTS are dumped in the same order, fields in the same order, etc.

This will enable comparing unique NFTs from a mono-service state file with one from a modularized service state file, where the actual representation at the platform level is different.

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

Dumps unique NFTs from a signed state file.

**Related issue(s)**:

Fixes #8255 

**Notes for reviewer**:

You can use the following command line (except providing your own signed state file):

```bash
./platform-sdk/swirlds-cli/pcli.sh \
   --memory 40 \
   --ignore-jars \
   --load hedera-node/data/lib \
   --load hedera-node/cli-clients/build/libs \
   --cli com.hedera.services.cli.signedstate \
   signed-state  -f /Users/davidbakin/StatesStreams/mainnet-2023-07-19/state/2023-07-19.00.00/144535932/SignedState.swh --verbose \
   dump \
      uniques --unique ./unique.lst --summary
```

This produced a 262MB file from a mainnet state of 2023-07, >2M unique NFTs.

Start of the file looked like this:
```
=== 2155205 unique tokens (863628 owned by treasury accounts)
          0 null owners, 2094718 null or missing spenders

nftId,nftSerial,owner;spender;creationTime;metadata;prev;next
410224;1;0.0.410226;;1630615611.855620000;5E192EB3;0.0.0.0;0.0.0.0
428976;1;0.0.309540;;1631258839.930284000;"{""name"":""Xact Logo V1"",""creator"":""Xact Team"",""category"":""Art"",""supply"":2}";0.0.428976.2;0.0.0.0
428976;2;0.0.309540;;1631258839.930284000;"{""name"":""Xact Logo V1"",""creator"":""Xact Team"",""category"":""Art"",""supply"":2}";0.0.446449.10;0.0.428976.1
439270;1;0.0.439272;;1631634298.625166000;5D16BC75;0.0.0.0;0.0.0.0
439393;1;0.0.439395;;1631636441.91550000;28BC0F90;0.0.0.0;0.0.0.0
446388;1;;;1631803925.656168000;"{""name"":""Xact Lines"",""creator"":""Xact Team"",""category"":""Art"",""supply"":10}";0.0.0.0;0.0.0.0
```

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (manual)

Signed-off-by: David Bakin <117694041+david-bakin-sl@users.noreply.github.com>

### CODE REVIEW

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                    null != mut.getMetadata() ? mut.getMetadata() : EMPTY_BYTES,
+                    mut.getPrev(),
+                    mut.getNext());
+        }
+    }
+
+    void reportSummary(@NonNull final Writer writer, @NonNull final Map<UniqueNFTId, UniqueNFT> uniques) {
+        final var relatedEntityCounts = RelatedEntities.countRelatedEntities(uniques);
+        writer.writeln("=== %7d unique tokens".formatted(uniques.size()));
+        writer.writeln("    %7d null or missing owners, %7d null or missing spenders"
```
Comment (tinker-michaelj) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java:159]: Per below (`0.0.0` means treasury-owned), `null or missing owners` might be better put as `held by treasuries` or something like that.

**Review by tinker-michaelj — DISMISSED** [review]
This all looks great to me! Have you considered a command to validate the owned NFT "linked lists" in state?

**Comment by tinker-michaelj** [comment]
Oh, but CI checks never ran for some reason. 🤔

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                        uniques.size() - relatedEntityCounts.spenders()));
+        writer.writeln("");
+    }
+
+    record RelatedEntities(long owners, long spenders) {
+        @NonNull
+        static RelatedEntities countRelatedEntities(@NonNull final Map<UniqueNFTId, UniqueNFT> uniques) {
+            final var cs = new long[2];
+            uniques.values().forEach(unique -> {
+                if (null != unique.owner && !unique.owner.equals(EntityId.MISSING_ENTITY_ID)) cs[0]++;
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java:174]: Summary at start of file now looks like:

```
=== 2155205 unique tokens (863628 owned by treasury accounts)
          0 null owners, 2094718 null or missing spenders

nftId,nftSerial,owner;spender;creationTime;metadata;prev;next
410224;1;0.0.410226;;1630615611.855620000;5E192EB3;0.0.0.0;0.0.0.0
```

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java (hunk 1 above)
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpUniqueTokensSubcommand.java]: See above comment for how it looks now.

**Comment by david-bakin-sl** [comment]
> This all looks great to me! Have you considered a command to validate the owned NFT "linked lists" in state?

Great idea! I'll do this shortly (but not in this PR).

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Pull request merged**